Functionnal test

AdrienMQ opened the pull request:
These patches add the functionnal tests to the parameter-framework.
They also dissociate the make install and the make test.

dawagner commented on CMakeLists.txt:
```diff
@@ -38,17 +38,22 @@ project(parameter-framework)
 endif()
 
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
```
duplicate

AdrienMQ replied:
done

dawagner commented on bindings/c/CMakeLists.txt:
```diff
@@ -56,5 +56,8 @@ if(BUILD_TESTING)
+    add_test(NAME cparameterUnitTest
+             COMMAND sh -c "PATH=${CMAKE_RUNTIME_OUTPUT_DIRECTORY}:$PATH \\
+                            LD_LIBRARY_PATH=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}:$LD_LIBRARY_PATH \\
+                            cparameterUnitTest")
```
use set_property for setting the test's environment

AdrienMQ replied:
done

dawagner commented on test/test-platform/CMakeLists.txt:
```diff
@@ -26,18 +26,20 @@
-target_link_libraries(test-platform parameter remote-processor)
-
-install(TARGETS test-platform RUNTIME DESTINATION bin)
+if (BUILD_TESTING)
```
don't condition the build of test-platform: it can be used for something else than testing the parameter-framework

AdrienMQ replied:
done

dawagner commented on test/test-subsystem/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+        "${PROJECT_SOURCE_DIR}/parameter"
+        "${PROJECT_SOURCE_DIR}/parameter/include") 
+    
+    target_link_libraries(test-subsystem parameter remote-processor)
```
I don't think libtest-subsystem needs to link against libremote-processor.

AdrienMQ replied:
conserve only the parameter and remove remove-process

dawagner commented on test/test-subsystem/TESTSubsystem.cpp:
```diff
@@ -0,0 +1,111 @@
+// Susbsystem sanity health
+bool CTESTSubsystem::isAlive() const
+{
+    return read(std::string(gacFwNamePropName) + "/isAlive") == "true";
```
`gacFwNamePropName` could be `NULL` which would cause an unmanaged exception. Please assert that `gacFwNamePropName != NULL`

AdrienMQ replied:
done

dawagner commented on test/test-subsystem/TESTSubsystem.cpp:
```diff
@@ -0,0 +1,111 @@
+// Resynchronization after subsystem restart needed
+bool CTESTSubsystem::needResync(bool bClear)
+{
+    std::string strNeedResyncFile = std::string(gacFwNamePropName) + "/needResync";
```
ditto

AdrienMQ replied:
done

dawagner commented on test/test-subsystem/TESTSubsystem.h:
```diff
@@ -0,0 +1,50 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
2011-2015

dawagner commented on test/test-subsystem/TESTSubsystemBinary.cpp:
```diff
@@ -0,0 +1,64 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/test-subsystem/TESTSubsystemBinary.h:
```diff
@@ -0,0 +1,47 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/test-subsystem/TESTSubsystemBuilder.cpp:
```diff
@@ -0,0 +1,42 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/test-subsystem/TESTSubsystemObject.cpp:
```diff
@@ -0,0 +1,150 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/test-subsystem/TESTSubsystemObject.h:
```diff
@@ -0,0 +1,59 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/test-subsystem/TESTSubsystemString.cpp:
```diff
@@ -0,0 +1,50 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/test-subsystem/TESTSubsystemString.h:
```diff
@@ -0,0 +1,45 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
ditto

dawagner commented on test/functionnal-tests/ACTCampaignEngine.py:
```diff
@@ -0,0 +1,90 @@
+    finally:
+        os.environ["PFW_FILESYSTEM"] = ""
+        os.environ["PFW_TEST_TOOLS"] = ""
+        os.environ["PFW_TEST_CONFIGURATION"] = ""
```
After this, the environment will still contains these variables (they will contain an empty string but still exist nonetheless). Do this instead: `del os.environ["PFW_FILESYSTEM"]`

AdrienMQ replied:
Remove try/finally.
Env vars are deleted when the script end.

dawagner commented:
some copyright years still need to be updated

dawagner commented:
Please clean all whitespaces by running `git rebase -f --whitespace=fix v2.4.2`

dawagner commented on bindings/c/CMakeLists.txt:
```diff
@@ -56,5 +56,11 @@ if(BUILD_TESTING)
     target_link_libraries(cparameterUnitTest cparameter)
-    add_test(cparameterUnitTest cparameterUnitTest)
+    add_test(NAME cparameterUnitTest
+             COMMAND cparameterUnitTest)
```
is this change really needed ?

dawagner commented on test/functionnal-tests/ACTCampaignEngine.py:
```diff
@@ -0,0 +1,79 @@
+import unittest
+import argparse
+import shutil
+import tempfile
```
argparse, shutil and tempfile aren't used.

dawagner commented on test/functionnal-tests/ACTCampaignEngine.py:
```diff
@@ -0,0 +1,79 @@
+
+def main():
+
+    os.environ["PFW_FILESYSTEM"] = os.environ["PFW_FS"]
```
I'm worried that 2 consecutive runs of the functional tests will reuse the same files and the second run might be poisoned. I'm not sure this is a problem.

AdrienMQ replied:
Now, we remove the test-parameter folder at the end of the tests (we use shutils.rmtree)

dawagner commented on test/functionnal-tests/CMakeLists.txt:
```diff
@@ -0,0 +1,58 @@
+
+    add_test(NAME functionnal-test 
+             WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} 
+             COMMAND ${python2} ACTCampaignEngine.py)
```
When this command is run, a lot of ".pyc" files are generated in the sources. It seems important to me that this is avoided.

AdrienMQ replied:
Done
But in counterparty, it seems that these files represent a time gain while executing tests. .pyc file are loaded faster than .py files.

dawagner commented:
See a77c5af82c621b7082dd22c60af137c84a14b5a1 for how to avoid valgrind/memcheck to be run on the functional tests; which takes a lot of time and is useless.

dawagner commented on test/functionnal-tests/README.md:
```diff
@@ -0,0 +1,15 @@
+#FUNCTIONNAL TEST
```
no need for ALL CAPS

dawagner commented on test/functionnal-tests/README.md:
```diff
@@ -0,0 +1,15 @@
+#FUNCTIONNAL TEST
+
+##What ?
+Functionnal tests are unit tests that ensure the functionment of the parameter-framework.
```
be careful: functionnal test are *not* unit tests. They are conceptually different

dawagner commented on test/functionnal-tests/README.md:
```diff
@@ -0,0 +1,15 @@
+
+##Practical
+
+Tests can be lunch with the 'make test' command.
```
launched

dawagner commented on test/functionnal-tests/README.md:
```diff
@@ -0,0 +1,15 @@
+##Practical
+
+Tests can be lunch with the 'make test' command.
+The test flag in cmake is set to 'ON' by default.
```
not relevant.

dawagner commented on test/functionnal-tests/README.md:
```diff
@@ -0,0 +1,15 @@
+The test flag in cmake is set to 'ON' by default.
+
+Environment vars are set in the cmake.
+Then we lunch ACTCampaignEngine.py which execute all the tests one by one.
```
launch

dawagner commented on test/test-subsystem/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+
+    set(PFW_FS "${CMAKE_BINARY_DIR}/tmp/test-parameters/")
+    file(WRITE "${PFW_FS}/isAlive" "true")
+    file(WRITE "${PFW_FS}/needResync" "false")
```
I believe these 3 commands should be part of the functional test's makefile; not here.

AdrienMQ replied:
done

dawagner commented:
`s/functionnal/functional/` (everywhere)

dawagner commented on test/functional-tests/ACTCampaignEngine.py:
```diff
@@ -0,0 +1,94 @@
+    sys.exit(not res)
+
+if __name__ == "__main__":
+    main()
```
1. When creating file paths, use os.path.join instead of "+" betwteen strings, which may be wrong (e.g. "/some/path" + "someFileName" == /some/pathsomeFileName" whereas os.path.join("/some/path", "someFileName") == "/some/path/someFileName")
2. remove the commented line
3. remove surrounding spaces around ˋos.environ["PFW_FS"]ˋ

dawagner commented on test/functional-tests/CMakeLists.txt:
```diff
@@ -0,0 +1,58 @@
+            LD_LIBRARY_PATH=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}:$ENV{LD_LIBRARY_PATH}
+            PFW_FS=${PFW_FS})
+
+endif()
```
you could set the PFW_FILESYSTEM env var directly

dawagner commented:
did you run the "git rebase -f --whitespace=fix" command ? (I can't check by myself right now).

dawagner commented:
The "for" loop in ACTCampaignEngine has several issues:

- if any exception is raised in testRunner, it will not be executed (need for a try...finally block)
- it isn't documented
- I'm not sure why you don't remove the whole PFW_FS directory.
-

AdrienMQ commented:
I just seen your comments. 
So, I think I have taken all your comments into account.

=> Removed some includes dir in test-subsystem
=> Run "git rebase -f --whitespace=fix"
=> For the copyright, it's done but we need to update all of them (including in python files) within a new patch.
=> I didn't remove the wall pfw path because of the xml temporary file which is needed to relaunch the test. I have implemented the try finally for the removal.


I have also changed the name of the environment variables. PFW_FS become PFW_ROOT and PFW_FILESYSTEM become PFW_RESULT (PFW_RESULT=PFW_ROOT/result).
Now, we have the xml directory in PFW_ROOT and I redirect the output on PFW_RESULT.
I remove PFW_RESULT at the end of the test but we conserve the xml files to have the availability to relaunch the tests.

Now, I have a persistant error which is : ParameterMgr: Unable to start remote processor server on port 5000. It seems to be a random error. I'm working on this, I'll push my changes when this will be fixed.

Thank you for your review. :+1:

AdrienMQ commented:
:+1:

dawagner commented:
I see the following log in several places in the functional test logs:

```
Please provide absolute path
```

This seems like an error but the test cases pass nonetheless; there must be an issue with these tests cases (probaly tied to the fact that `PFW_TEST_TOOLS = "."`.

Another error:

```
3: test_Restore_Config_Error (Domains.tDomain_Configuration_Backup.TestCases) ... Command not found!
```

here, the test is supposed to be checking that an error is returned when there is a missing argument to the command, *not if the command name is wrong* (indeed, the `RestoreConfiguration` command is called instead of `restoreConfiguration`

Please check all errors; especially the test cases that used to fail.

I've also left some comments on f851cb63db821bd9d9e5a403bfc7904cc207ae85

You're almost there :)

AdrienMQ commented:
I have fixed your comments bit it seems to be some errors related in the log : tFunctions_ListingFunction, tBit_Block ....

AdrienMQ commented:
Update last modifications.
According to J. Clero, some test are not fixed and must fail. With the changes brought by this pull request, it is hard to verify manually.

AdrienMQ commented:
Not a diff between local and remote version but a bug in the visualisation?
